cpp/src/shuffler/shuffler.cpp

```diff
@@ -282,9 +422,18 @@ class Shuffler::Progress {
                 shuffler_.statistics_->add_bytes_stat(
                     "shuffle-payload-send", chunk.gpu_data->size
                 );
-                fire_and_forget_.push_back(
-                    shuffler_.comm_->send(std::move(chunk.gpu_data), src, gpu_data_tag)
-                );
+                try {
+                    fire_and_forget_.push_back(shuffler_.comm_->send(
+                        std::move(chunk.gpu_data), src, gpu_data_tag
+                    ));
+                } catch (std::logic_error& e) {
+                    RAPIDSMPF_EXPECTS(
+                        outgoing_chunks_
+                            .insert({ready_for_data_msg.cid, std::move(chunk)})
```

Comment:
this `chunk` doesn't have any `gpu_data` now, does it? 

I personally dislike the blanket exception catch here. This catches non-synchronization exceptions like [these](https://github.com/rapidsai/rapidsmpf/blob/branch-25.06/cpp/src/communicator/mpi.cpp#L119-L122) as well. 
I suggest using a return flag, like `pair<bool, unique_ptr<Buffer>>` as the output from `comm::send`. This will allow us to throw and exit as well as return with failure gracefully (and revisit later).

Reply:
Good point. However, returning this return type will be very clunky to manage, in fact the current return type is a `Future`, which would make doing something like what you suggested even more clunky. We would need something like `pair<bool, unique_ptr<Future>>` and the user would be responsible to check the value of the `bool` and then if the send failed extract the `Buffer` from `Future` to then continue, at this point it is much simpler for `comm::send` and the caller to just add a `Buffer::is_copy_complete()` before calling `comm::send`.

Reply:
As per https://github.com/rapidsai/rapidsmpf/pull/225#discussion_r2059099434, I'll revert those changes and update the docstring to clarify the contract.